{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "HW4 Tardigrades."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Obtaining data \n",
    "\n",
    "1.1 Genomic: `wget -o data/Tardi_assebled_genome.fna.gz http://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/001/949/185/GCA_001949185.1_Rvar_4.0/GCA_001949185.1_Rvar_4.0_genomic.fna.gz` \\\n",
    "1.2 AUGUSTUS results: https://drive.google.com/file/d/1hCEywBlqNzTrIpQsZTVuZk1S9qKzqQAq/view?usp=sharing \\\n",
    "1.3 GFF: https://drive.google.com/file/d/12ShwrgLkvJIYQV2p1UlXklmxSOOxyxj4/view?usp=sharing \\\n",
    "1.4 getAnnoFasta.pl script: http://augustus.gobics.de/binaries/scripts/getAnnoFasta.pl \\\n",
    "1.5 Tandem Mass Spectrometry results: https://disk.yandex.ru/d/xJqQMGX77Xueqg \\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Get number of obtained proteins \n",
    "\n",
    "`grep \">\" data/augustus.whole.aa | wc -l`\n",
    "-> 16435 \\\n",
    "2.1 To narrow down the amount of genes we can look for those which take their place in a nucleus, thus having a NLS (Nuclear Localization Signals). Classical NLS (cNLS): This type of signal is recognized by importin α and importin β proteins. The classical NLS is usually a short stretch of basic amino acids, such as lysine or arginine. The consensus sequence for classical NLS is typically (K/R)4-5(K/R)X(K/R)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Do a local alignment-based search \n",
    "\n",
    "3.1 *SP*Option a*SP* \\\n",
    "`conda install -c bioconda blast` \\\n",
    "`makeblastdb -in data/augustus.whole.aa -dbtype prot  -out data/blastp_tardigrade_protein_db` \\\n",
    "`blastp -db data/blastp_tardigrade_protein_db -query data/peptides.fa -outfmt 6 -out data/blastp_tardigrade_filtered_proteins` \\\n",
    "Column names format 6: qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore \\\n",
    "3.2 *SP*Option b*SP* \\\n",
    "`conda install -c bioconda diamond` \\\n",
    "`diamond makedb --in data/augustus.whole.aa  --db data/diamond_tardigrade_protein_db` \\\n",
    "`diamond blastp -d data/diamond_tardigrade_protein_db.dmnd -q data/peptides.fa  -f 6 -o data/diamond_tardigrade_filtered_proteins --very-sensitive` \\\n",
    "Used option b. From TMS created a local database with peptides. Then blasted against fasta file with all proteins."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Localization prediction \n",
    "\n",
    "Get fasta's for filtered genes \\\n",
    "for 3a:\n",
    "- unique gene names\n",
    "`awk -F'\\t' '{print $2}' data/blastp_tardigrade_filtered_proteins | sort -u | wc -l`\n",
    "`awk -F'\\t' '{print $2}' data/blastp_tardigrade_filtered_proteins | sort -u > data/blastp_unique_genes` \\\n",
    "- get fasta sequences from the original fasta for every unique gene name\n",
    "`samtools faidx data/augustus.whole.aa`\n",
    "`xargs samtools faidx data/augustus.whole.aa < data/blastp_unique_genes > data/blastp_selected_sequences.fasta` \\\n",
    "- get all gene names where prediction is SP\n",
    "`awk -F'\\t' '$2 == \"SP\" {print $1}' data/blastp_targetP_output_protein_type > data/blastp_targetP_selected_genes`\n",
    "- get sequences for them\n",
    "`xargs samtools faidx data/augustus.whole.aa < data/blastp_targetP_selected_genes > data/blastp_targetP_selected_genes.fasta`\n",
    "\n",
    "for 3b:\n",
    "4a. WoLF PSORT  (https://wolfpsort.hgc.jp/) predicts the subcellular localization of proteins. PSORT predicts cellular localization of proteins based on the presence of a signal peptide on their N-terminus. \\\n",
    "Results: \\\n",
    "g4106.t1 details E.R.: 14.5, E.R._golg: 9.5, extr: 7, golg: 3.5, lyso: 3, pero: 2, plas: 1, mito: 1 \\\n",
    "g12510.t1 details plas: 29, cyto: 3 \\\n",
    "(see pdfs) \\\n",
    "4b. TargetP (https://services.healthtech.dtu.dk/service.php?TargetP-2.0) also predicts the subcellular localization of eukaryotic proteins. The location assignment is based on the predicted presence of any of the N-terminal presequences: chloroplast transit peptide (cTP), mitochondrial targeting peptide (mTP) or secretory pathway signal peptide (SP). \\\n",
    "Results: \\\n",
    "(see zip and pdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. BLAST\n",
    "\n",
    "BLAST (https://blast.ncbi.nlm.nih.gov/Blast.cgi) protein sequences against the “UniProtKB/Swiss-Prot” database.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. PFAM\n",
    "\n",
    "We will use HMMER (web-version, https://www.ebi.ac.uk/Tools/hmmer/) to search our protein sequences against a collection of profile-HMMs for different protein domains and motifs.\n",
    "Select “Search” and then select the suitable tool (in our case it is hmmscan), select the Pfam database, and submit your proteins.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge tables with WoLF PSORT and TargetP results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Prediction</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>SP</th>\n",
       "      <th>mTP</th>\n",
       "      <th>CS Position</th>\n",
       "      <th>Contents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g10513.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 20, cyto_nucl: 14.5, cyto: 7, extr: 3, E.R.: 1, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g10514.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999543</td>\n",
       "      <td>0.000349</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 19, cyto_nucl: 15, cyto: 9, extr: 3, mito: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g11320.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000184</td>\n",
       "      <td>0.999816</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CS pos: 20-21. AYS-AG. Pr: 0.7236</td>\n",
       "      <td>plas: 24.5, extr_plas: 16, extr: 6.5, lyso: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g11513.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999434</td>\n",
       "      <td>0.000401</td>\n",
       "      <td>0.000164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cyto: 17, cyto_nucl: 12.8333, cyto_mito: 9.83333, nucl: 7.5, E.R.: 3, mito: 1.5, plas: 1, pero: 1, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g11806.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.998977</td>\n",
       "      <td>0.000887</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 18, cyto_nucl: 11.8333, mito: 5, extr: 4, cyto: 3.5, cyto_pero: 2.66667, cysk_plas: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>g11960.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>g12388.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000490</td>\n",
       "      <td>0.999481</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>CS pos: 16-17. ASA-SS. Pr: 0.6485</td>\n",
       "      <td>extr: 25, plas: 4, mito: 2, lyso: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>g12510.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999738</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plas: 29, cyto: 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>g12562.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.999923</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>CS pos: 16-17. SYA-AN. Pr: 0.7910</td>\n",
       "      <td>extr: 30, lyso: 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>g1285.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.003029</td>\n",
       "      <td>0.996798</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>CS pos: 16-17. ASA-TS. Pr: 0.7127</td>\n",
       "      <td>extr: 25, plas: 5, mito: 1, lyso: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>g13530.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.116007</td>\n",
       "      <td>0.883840</td>\n",
       "      <td>0.000153</td>\n",
       "      <td>CS pos: 19-20. TIP-FT. Pr: 0.3552</td>\n",
       "      <td>extr: 13, nucl: 6.5, lyso: 5, cyto_nucl: 4.5, plas: 3, E.R.: 3, cyto: 1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>g14472.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 28, plas: 2, cyto: 1, cysk: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>g15153.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.999986</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CS pos: 16-17. AYA-AN. Pr: 0.8378</td>\n",
       "      <td>extr: 32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>g15484.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999980</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 17.5, cyto_nucl: 15.3333, cyto: 12, cyto_mito: 6.83333, plas: 1, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>g16318.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.997047</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 20.5, cyto_nucl: 13, extr: 5, cyto: 4.5, E.R.: 1, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>g16368.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.996693</td>\n",
       "      <td>0.003307</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 20.5, cyto_nucl: 13, extr: 5, cyto: 4.5, E.R.: 1, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>g2203.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999869</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plas: 29, nucl: 2, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>g3428.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999903</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mito: 18, cyto: 11, extr: 2, nucl: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>g3679.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.001755</td>\n",
       "      <td>0.998023</td>\n",
       "      <td>0.000222</td>\n",
       "      <td>CS pos: 18-19. TFA-AR. Pr: 0.5523</td>\n",
       "      <td>extr: 26, mito: 2, lyso: 2, plas: 1, E.R.: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>g4106.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.729658</td>\n",
       "      <td>0.266917</td>\n",
       "      <td>0.003425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E.R.: 14.5, E.R._golg: 9.5, extr: 7, golg: 3.5, lyso: 3, pero: 2, plas: 1, mito: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>g4970.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plas: 32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>g5237.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999545</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.000111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plas: 24, mito: 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>g5443.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.952853</td>\n",
       "      <td>0.043784</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>NaN</td>\n",
       "      <td>extr: 28, nucl: 3, cyto: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>g5467.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>0.999845</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>CS pos: 16-17. ASA-GS. Pr: 0.6543</td>\n",
       "      <td>extr: 27, plas: 4, mito: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>g5502.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.001134</td>\n",
       "      <td>0.998823</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>CS pos: 16-17. ASA-GS. Pr: 0.6833</td>\n",
       "      <td>extr: 31, lyso: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>g5503.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.001222</td>\n",
       "      <td>0.998720</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>CS pos: 16-17. ASA-GS. Pr: 0.6833</td>\n",
       "      <td>extr: 29, plas: 1, mito: 1, lyso: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>g5510.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999108</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>0.000876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plas: 23, mito: 7, E.R.: 1, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>g5616.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.999933</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>CS pos: 16-17. ACA-AN. Pr: 0.5270</td>\n",
       "      <td>extr: 31, mito: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>g5641.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.999869</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>CS pos: 16-17. ACA-AS. Pr: 0.4873</td>\n",
       "      <td>extr: 31, lyso: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>g5927.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 30.5, cyto_nucl: 16.5, cyto: 1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>g702.t1</td>\n",
       "      <td>SP</td>\n",
       "      <td>0.000347</td>\n",
       "      <td>0.999652</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>CS pos: 16-17. ALA-AN. Pr: 0.8153</td>\n",
       "      <td>extr: 29, plas: 2, lyso: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>g7861.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999975</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 16, cyto_nucl: 14, cyto: 8, plas: 5, pero: 1, cysk: 1, golg: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>g8100.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999955</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 16.5, cyto_nucl: 12.5, cyto: 7.5, plas: 5, extr: 2, E.R.: 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>g8312.t1</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>0.999930</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucl: 15.5, cyto_nucl: 15.5, cyto: 12.5, mito: 2, plas: 1, golg: 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene Prediction     OTHER        SP       mTP  \\\n",
       "0   g10513.t1      OTHER  0.999999  0.000001  0.000000   \n",
       "1   g10514.t1      OTHER  0.999543  0.000349  0.000107   \n",
       "2   g11320.t1         SP  0.000184  0.999816  0.000000   \n",
       "3   g11513.t1      OTHER  0.999434  0.000401  0.000164   \n",
       "4   g11806.t1      OTHER  0.998977  0.000887  0.000136   \n",
       "5   g11960.t1      OTHER  0.999996  0.000002  0.000002   \n",
       "6   g12388.t1         SP  0.000490  0.999481  0.000029   \n",
       "7   g12510.t1      OTHER  0.999738  0.000099  0.000163   \n",
       "8   g12562.t1         SP  0.000076  0.999923  0.000001   \n",
       "9    g1285.t1         SP  0.003029  0.996798  0.000173   \n",
       "10  g13530.t1         SP  0.116007  0.883840  0.000153   \n",
       "11  g14472.t1      OTHER  0.999999  0.000001  0.000000   \n",
       "12  g15153.t1         SP  0.000014  0.999986  0.000000   \n",
       "13  g15484.t1      OTHER  0.999980  0.000010  0.000010   \n",
       "14  g16318.t1      OTHER  0.997047  0.002953  0.000000   \n",
       "15  g16368.t1      OTHER  0.996693  0.003307  0.000000   \n",
       "16   g2203.t1      OTHER  0.999869  0.000031  0.000100   \n",
       "17   g3428.t1      OTHER  0.999903  0.000033  0.000064   \n",
       "18   g3679.t1         SP  0.001755  0.998023  0.000222   \n",
       "19   g4106.t1      OTHER  0.729658  0.266917  0.003425   \n",
       "20   g4970.t1      OTHER  0.999996  0.000003  0.000001   \n",
       "21   g5237.t1      OTHER  0.999545  0.000345  0.000111   \n",
       "22   g5443.t1      OTHER  0.952853  0.043784  0.003363   \n",
       "23   g5467.t1         SP  0.000096  0.999845  0.000059   \n",
       "24   g5502.t1         SP  0.001134  0.998823  0.000043   \n",
       "25   g5503.t1         SP  0.001222  0.998720  0.000058   \n",
       "26   g5510.t1      OTHER  0.999108  0.000016  0.000876   \n",
       "27   g5616.t1         SP  0.000067  0.999933  0.000000   \n",
       "28   g5641.t1         SP  0.000130  0.999869  0.000001   \n",
       "29   g5927.t1      OTHER  0.999995  0.000001  0.000004   \n",
       "30    g702.t1         SP  0.000347  0.999652  0.000001   \n",
       "31   g7861.t1      OTHER  0.999975  0.000004  0.000022   \n",
       "32   g8100.t1      OTHER  0.999955  0.000024  0.000021   \n",
       "33   g8312.t1      OTHER  0.999930  0.000065  0.000004   \n",
       "\n",
       "                          CS Position  \\\n",
       "0                                 NaN   \n",
       "1                                 NaN   \n",
       "2   CS pos: 20-21. AYS-AG. Pr: 0.7236   \n",
       "3                                 NaN   \n",
       "4                                 NaN   \n",
       "5                                 NaN   \n",
       "6   CS pos: 16-17. ASA-SS. Pr: 0.6485   \n",
       "7                                 NaN   \n",
       "8   CS pos: 16-17. SYA-AN. Pr: 0.7910   \n",
       "9   CS pos: 16-17. ASA-TS. Pr: 0.7127   \n",
       "10  CS pos: 19-20. TIP-FT. Pr: 0.3552   \n",
       "11                                NaN   \n",
       "12  CS pos: 16-17. AYA-AN. Pr: 0.8378   \n",
       "13                                NaN   \n",
       "14                                NaN   \n",
       "15                                NaN   \n",
       "16                                NaN   \n",
       "17                                NaN   \n",
       "18  CS pos: 18-19. TFA-AR. Pr: 0.5523   \n",
       "19                                NaN   \n",
       "20                                NaN   \n",
       "21                                NaN   \n",
       "22                                NaN   \n",
       "23  CS pos: 16-17. ASA-GS. Pr: 0.6543   \n",
       "24  CS pos: 16-17. ASA-GS. Pr: 0.6833   \n",
       "25  CS pos: 16-17. ASA-GS. Pr: 0.6833   \n",
       "26                                NaN   \n",
       "27  CS pos: 16-17. ACA-AN. Pr: 0.5270   \n",
       "28  CS pos: 16-17. ACA-AS. Pr: 0.4873   \n",
       "29                                NaN   \n",
       "30  CS pos: 16-17. ALA-AN. Pr: 0.8153   \n",
       "31                                NaN   \n",
       "32                                NaN   \n",
       "33                                NaN   \n",
       "\n",
       "                                                                                                      Contents  \n",
       "0                                                nucl: 20, cyto_nucl: 14.5, cyto: 7, extr: 3, E.R.: 1, golg: 1  \n",
       "1                                                           nucl: 19, cyto_nucl: 15, cyto: 9, extr: 3, mito: 1  \n",
       "2                                                                plas: 24.5, extr_plas: 16, extr: 6.5, lyso: 1  \n",
       "3   cyto: 17, cyto_nucl: 12.8333, cyto_mito: 9.83333, nucl: 7.5, E.R.: 3, mito: 1.5, plas: 1, pero: 1, golg: 1  \n",
       "4                  nucl: 18, cyto_nucl: 11.8333, mito: 5, extr: 4, cyto: 3.5, cyto_pero: 2.66667, cysk_plas: 1  \n",
       "5                                                                                                     nucl: 32  \n",
       "6                                                                          extr: 25, plas: 4, mito: 2, lyso: 1  \n",
       "7                                                                                            plas: 29, cyto: 3  \n",
       "8                                                                                            extr: 30, lyso: 2  \n",
       "9                                                                          extr: 25, plas: 5, mito: 1, lyso: 1  \n",
       "10                                   extr: 13, nucl: 6.5, lyso: 5, cyto_nucl: 4.5, plas: 3, E.R.: 3, cyto: 1.5  \n",
       "11                                                                         nucl: 28, plas: 2, cyto: 1, cysk: 1  \n",
       "12                                                                                                    extr: 32  \n",
       "13                              nucl: 17.5, cyto_nucl: 15.3333, cyto: 12, cyto_mito: 6.83333, plas: 1, golg: 1  \n",
       "14                                             nucl: 20.5, cyto_nucl: 13, extr: 5, cyto: 4.5, E.R.: 1, golg: 1  \n",
       "15                                             nucl: 20.5, cyto_nucl: 13, extr: 5, cyto: 4.5, E.R.: 1, golg: 1  \n",
       "16                                                                                  plas: 29, nucl: 2, golg: 1  \n",
       "17                                                                        mito: 18, cyto: 11, extr: 2, nucl: 1  \n",
       "18                                                                extr: 26, mito: 2, lyso: 2, plas: 1, E.R.: 1  \n",
       "19                          E.R.: 14.5, E.R._golg: 9.5, extr: 7, golg: 3.5, lyso: 3, pero: 2, plas: 1, mito: 1  \n",
       "20                                                                                                    plas: 32  \n",
       "21                                                                                           plas: 24, mito: 8  \n",
       "22                                                                                  extr: 28, nucl: 3, cyto: 1  \n",
       "23                                                                                  extr: 27, plas: 4, mito: 1  \n",
       "24                                                                                           extr: 31, lyso: 1  \n",
       "25                                                                         extr: 29, plas: 1, mito: 1, lyso: 1  \n",
       "26                                                                         plas: 23, mito: 7, E.R.: 1, golg: 1  \n",
       "27                                                                                           extr: 31, mito: 1  \n",
       "28                                                                                           extr: 31, lyso: 1  \n",
       "29                                                                      nucl: 30.5, cyto_nucl: 16.5, cyto: 1.5  \n",
       "30                                                                                  extr: 29, plas: 2, lyso: 1  \n",
       "31                                        nucl: 16, cyto_nucl: 14, cyto: 8, plas: 5, pero: 1, cysk: 1, golg: 1  \n",
       "32                                           nucl: 16.5, cyto_nucl: 12.5, cyto: 7.5, plas: 5, extr: 2, E.R.: 1  \n",
       "33                                          nucl: 15.5, cyto_nucl: 15.5, cyto: 12.5, mito: 2, plas: 1, golg: 1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{llrrrll}\n",
      "\\toprule\n",
      "     Gene & Prediction &    OTHER &       SP &      mTP &                       CS Position &                                                                                                   Contents \\\\\n",
      "\\midrule\n",
      "g10513.t1 &      OTHER & 0.999999 & 0.000001 & 0.000000 &                               NaN &                                              nucl: 20, cyto_nucl: 14.5, cyto: 7, extr: 3, E.R.: 1, golg: 1 \\\\\n",
      "g10514.t1 &      OTHER & 0.999543 & 0.000349 & 0.000107 &                               NaN &                                                         nucl: 19, cyto_nucl: 15, cyto: 9, extr: 3, mito: 1 \\\\\n",
      "g11320.t1 &         SP & 0.000184 & 0.999816 & 0.000000 & CS pos: 20-21. AYS-AG. Pr: 0.7236 &                                                              plas: 24.5, extr_plas: 16, extr: 6.5, lyso: 1 \\\\\n",
      "g11513.t1 &      OTHER & 0.999434 & 0.000401 & 0.000164 &                               NaN & cyto: 17, cyto_nucl: 12.8333, cyto_mito: 9.83333, nucl: 7.5, E.R.: 3, mito: 1.5, plas: 1, pero: 1, golg: 1 \\\\\n",
      "g11806.t1 &      OTHER & 0.998977 & 0.000887 & 0.000136 &                               NaN &                nucl: 18, cyto_nucl: 11.8333, mito: 5, extr: 4, cyto: 3.5, cyto_pero: 2.66667, cysk_plas: 1 \\\\\n",
      "g11960.t1 &      OTHER & 0.999996 & 0.000002 & 0.000002 &                               NaN &                                                                                                   nucl: 32 \\\\\n",
      "g12388.t1 &         SP & 0.000490 & 0.999481 & 0.000029 & CS pos: 16-17. ASA-SS. Pr: 0.6485 &                                                                        extr: 25, plas: 4, mito: 2, lyso: 1 \\\\\n",
      "g12510.t1 &      OTHER & 0.999738 & 0.000099 & 0.000163 &                               NaN &                                                                                          plas: 29, cyto: 3 \\\\\n",
      "g12562.t1 &         SP & 0.000076 & 0.999923 & 0.000001 & CS pos: 16-17. SYA-AN. Pr: 0.7910 &                                                                                          extr: 30, lyso: 2 \\\\\n",
      " g1285.t1 &         SP & 0.003029 & 0.996798 & 0.000173 & CS pos: 16-17. ASA-TS. Pr: 0.7127 &                                                                        extr: 25, plas: 5, mito: 1, lyso: 1 \\\\\n",
      "g13530.t1 &         SP & 0.116007 & 0.883840 & 0.000153 & CS pos: 19-20. TIP-FT. Pr: 0.3552 &                                  extr: 13, nucl: 6.5, lyso: 5, cyto_nucl: 4.5, plas: 3, E.R.: 3, cyto: 1.5 \\\\\n",
      "g14472.t1 &      OTHER & 0.999999 & 0.000001 & 0.000000 &                               NaN &                                                                        nucl: 28, plas: 2, cyto: 1, cysk: 1 \\\\\n",
      "g15153.t1 &         SP & 0.000014 & 0.999986 & 0.000000 & CS pos: 16-17. AYA-AN. Pr: 0.8378 &                                                                                                   extr: 32 \\\\\n",
      "g15484.t1 &      OTHER & 0.999980 & 0.000010 & 0.000010 &                               NaN &                             nucl: 17.5, cyto_nucl: 15.3333, cyto: 12, cyto_mito: 6.83333, plas: 1, golg: 1 \\\\\n",
      "g16318.t1 &      OTHER & 0.997047 & 0.002953 & 0.000000 &                               NaN &                                            nucl: 20.5, cyto_nucl: 13, extr: 5, cyto: 4.5, E.R.: 1, golg: 1 \\\\\n",
      "g16368.t1 &      OTHER & 0.996693 & 0.003307 & 0.000000 &                               NaN &                                            nucl: 20.5, cyto_nucl: 13, extr: 5, cyto: 4.5, E.R.: 1, golg: 1 \\\\\n",
      " g2203.t1 &      OTHER & 0.999869 & 0.000031 & 0.000100 &                               NaN &                                                                                 plas: 29, nucl: 2, golg: 1 \\\\\n",
      " g3428.t1 &      OTHER & 0.999903 & 0.000033 & 0.000064 &                               NaN &                                                                       mito: 18, cyto: 11, extr: 2, nucl: 1 \\\\\n",
      " g3679.t1 &         SP & 0.001755 & 0.998023 & 0.000222 & CS pos: 18-19. TFA-AR. Pr: 0.5523 &                                                               extr: 26, mito: 2, lyso: 2, plas: 1, E.R.: 1 \\\\\n",
      " g4106.t1 &      OTHER & 0.729658 & 0.266917 & 0.003425 &                               NaN &                         E.R.: 14.5, E.R._golg: 9.5, extr: 7, golg: 3.5, lyso: 3, pero: 2, plas: 1, mito: 1 \\\\\n",
      " g4970.t1 &      OTHER & 0.999996 & 0.000003 & 0.000001 &                               NaN &                                                                                                   plas: 32 \\\\\n",
      " g5237.t1 &      OTHER & 0.999545 & 0.000345 & 0.000111 &                               NaN &                                                                                          plas: 24, mito: 8 \\\\\n",
      " g5443.t1 &      OTHER & 0.952853 & 0.043784 & 0.003363 &                               NaN &                                                                                 extr: 28, nucl: 3, cyto: 1 \\\\\n",
      " g5467.t1 &         SP & 0.000096 & 0.999845 & 0.000059 & CS pos: 16-17. ASA-GS. Pr: 0.6543 &                                                                                 extr: 27, plas: 4, mito: 1 \\\\\n",
      " g5502.t1 &         SP & 0.001134 & 0.998823 & 0.000043 & CS pos: 16-17. ASA-GS. Pr: 0.6833 &                                                                                          extr: 31, lyso: 1 \\\\\n",
      " g5503.t1 &         SP & 0.001222 & 0.998720 & 0.000058 & CS pos: 16-17. ASA-GS. Pr: 0.6833 &                                                                        extr: 29, plas: 1, mito: 1, lyso: 1 \\\\\n",
      " g5510.t1 &      OTHER & 0.999108 & 0.000016 & 0.000876 &                               NaN &                                                                        plas: 23, mito: 7, E.R.: 1, golg: 1 \\\\\n",
      " g5616.t1 &         SP & 0.000067 & 0.999933 & 0.000000 & CS pos: 16-17. ACA-AN. Pr: 0.5270 &                                                                                          extr: 31, mito: 1 \\\\\n",
      " g5641.t1 &         SP & 0.000130 & 0.999869 & 0.000001 & CS pos: 16-17. ACA-AS. Pr: 0.4873 &                                                                                          extr: 31, lyso: 1 \\\\\n",
      " g5927.t1 &      OTHER & 0.999995 & 0.000001 & 0.000004 &                               NaN &                                                                     nucl: 30.5, cyto_nucl: 16.5, cyto: 1.5 \\\\\n",
      "  g702.t1 &         SP & 0.000347 & 0.999652 & 0.000001 & CS pos: 16-17. ALA-AN. Pr: 0.8153 &                                                                                 extr: 29, plas: 2, lyso: 1 \\\\\n",
      " g7861.t1 &      OTHER & 0.999975 & 0.000004 & 0.000022 &                               NaN &                                       nucl: 16, cyto_nucl: 14, cyto: 8, plas: 5, pero: 1, cysk: 1, golg: 1 \\\\\n",
      " g8100.t1 &      OTHER & 0.999955 & 0.000024 & 0.000021 &                               NaN &                                          nucl: 16.5, cyto_nucl: 12.5, cyto: 7.5, plas: 5, extr: 2, E.R.: 1 \\\\\n",
      " g8312.t1 &      OTHER & 0.999930 & 0.000065 & 0.000004 &                               NaN &                                         nucl: 15.5, cyto_nucl: 15.5, cyto: 12.5, mito: 2, plas: 1, golg: 1 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Brave\\AppData\\Local\\Temp\\ipykernel_9460\\348150165.py:25: FutureWarning: In future versions `DataFrame.to_latex` is expected to utilise the base implementation of `Styler.to_latex` for formatting and rendering. The arguments signature may therefore change. It is recommended instead to use `DataFrame.style.to_latex` which also contains additional functionality.\n",
      "  latex_code = result_df.to_latex(index=False, escape=False)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "with open (os.path.join('data', 'blastp_WoLFPSORT_results'), 'r') as fp:\n",
    "    wolf_results = {}\n",
    "    file_content = fp.readlines()[0:]\n",
    "    for line in file_content:\n",
    "        line_content = line.strip().split(' ')\n",
    "        wolf_results[line_content[0]] = (' ').join(line_content[2:])\n",
    "\n",
    "wolf_results_df = pd.DataFrame(list(wolf_results.items()), columns=['Gene', 'Contents'])\n",
    "\n",
    "targetP_df = pd.read_csv(os.path.join('data', 'blastp_selected_sequences_TargetP_results'), sep='\\t', skiprows=1, header=0)\n",
    "targetP_df.rename(columns={targetP_df.columns[0]: 'Gene'}, inplace=True)\n",
    "# blast_results_df = pd.read_csv(os.path.join('data', 'blastp_selected_sequences_BLAST_results.csv'), \n",
    "#                                names=['query acc.ver', 'subject acc.ver', '% identity', 'alignment length', 'mismatches', 'gap opens', 'q. start', 'q. end', 's. start', 's. end', 'evalue', 'bit score', '% positives'])\n",
    "\n",
    "result_df = targetP_df.merge(wolf_results_df, how='inner', left_on='Gene', right_on='Gene')\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "display(result_df)\n",
    "\n",
    "# result_df['Contents'][0]\n",
    "\n",
    "latex_code = result_df.to_latex(index=False, escape=False)\n",
    "print(latex_code)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now parse BLAST results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Gene Query Acc.Ver Subject Acc.Ver  % Identity  Alignment Length  \\\n",
      "0    g10513.t1           NaN             NaN         NaN               NaN   \n",
      "1    g10514.t1           NaN             NaN         NaN               NaN   \n",
      "2    g11320.t1           NaN             NaN         NaN               NaN   \n",
      "3    g11513.t1     g11513.t1        Q32PH0.1      28.607             811.0   \n",
      "4    g11806.t1     g11513.t1        Q96Q05.2      27.039             932.0   \n",
      "..         ...           ...             ...         ...               ...   \n",
      "899        NaN      g8100.t1        O67791.1      24.000             300.0   \n",
      "900        NaN      g8100.t1        Q8FAG5.1      24.232             293.0   \n",
      "901        NaN      g8100.t1        P10497.2      38.462              52.0   \n",
      "902        NaN      g8100.t1        Q45499.1      20.379             211.0   \n",
      "903        NaN      g8100.t1        P46726.2      39.655              58.0   \n",
      "\n",
      "     Mismatches  Gap Opens  Q. Start  Q. End  S. Start  S. End       E-value  \\\n",
      "0           NaN        NaN       NaN     NaN       NaN     NaN           NaN   \n",
      "1           NaN        NaN       NaN     NaN       NaN     NaN           NaN   \n",
      "2           NaN        NaN       NaN     NaN       NaN     NaN           NaN   \n",
      "3         492.0       19.0     219.0   966.0     143.0   929.0  6.640000e-83   \n",
      "4         568.0       23.0     218.0  1060.0     142.0  1050.0  3.590000e-81   \n",
      "..          ...        ...       ...     ...       ...     ...           ...   \n",
      "899       146.0       11.0     463.0   748.0       3.0   234.0  1.000000e-03   \n",
      "900       156.0        9.0     467.0   754.0       1.0   232.0  1.000000e-03   \n",
      "901        32.0        0.0     701.0   752.0      15.0    66.0  1.000000e-03   \n",
      "902       125.0        7.0     552.0   752.0      62.0   239.0  2.000000e-03   \n",
      "903        35.0        0.0     695.0   752.0     184.0   241.0  3.300000e-02   \n",
      "\n",
      "     Bit Score  % Positives  \n",
      "0          NaN          NaN  \n",
      "1          NaN          NaN  \n",
      "2          NaN          NaN  \n",
      "3        296.0        48.21  \n",
      "4        291.0        45.60  \n",
      "..         ...          ...  \n",
      "899       45.8        36.00  \n",
      "900       45.4        39.25  \n",
      "901       42.4        51.92  \n",
      "902       44.7        41.23  \n",
      "903       41.2        53.45  \n",
      "\n",
      "[904 rows x 14 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Initialize variables to store data\n",
    "current_query = None\n",
    "hits_data = []\n",
    "\n",
    "# Open the file for reading\n",
    "with open(os.path.join('data', 'blastp_selected_sequences_BLAST_results'), 'r') as file:\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "\n",
    "        # Check if the line contains information about the Query\n",
    "        if line.startswith('# Query:'):\n",
    "            # If a new Query is encountered, save the previous Query and its hits\n",
    "            if current_query is not None:\n",
    "                # Append the hits data to the list\n",
    "                hits_data.append({'Gene': current_query, 'Hits': hits})\n",
    "            \n",
    "            # Extract the new Query\n",
    "            current_query = line.split(': ')[1]\n",
    "            hits = []  # Reset the hits for the new Query\n",
    "        \n",
    "        # Check if the line contains information about a hit\n",
    "        elif line.startswith('# Fields:'):\n",
    "            # Skip lines until the hits start\n",
    "            next(file)  # Move to the next line to get the hits\n",
    "\n",
    "        elif not line.startswith('#') and line:  # Skip empty lines and lines starting with #\n",
    "            # Split the line to extract hit information\n",
    "            hit_info = line.split('\\t')\n",
    "            \n",
    "            # Create a dictionary for the hit with all the fields\n",
    "            hit = {\n",
    "                'Query Acc.Ver': hit_info[0],\n",
    "                'Subject Acc.Ver': hit_info[1],\n",
    "                '% Identity': float(hit_info[2]),\n",
    "                'Alignment Length': int(hit_info[3]),\n",
    "                'Mismatches': int(hit_info[4]),\n",
    "                'Gap Opens': int(hit_info[5]),\n",
    "                'Q. Start': int(hit_info[6]),\n",
    "                'Q. End': int(hit_info[7]),\n",
    "                'S. Start': int(hit_info[8]),\n",
    "                'S. End': int(hit_info[9]),\n",
    "                'E-value': float(hit_info[10]),\n",
    "                'Bit Score': float(hit_info[11]),\n",
    "                '% Positives': float(hit_info[12])\n",
    "            }\n",
    "            \n",
    "            # Store the hit information\n",
    "            hits.append(hit)\n",
    "\n",
    "# Create a DataFrame for the collected hits data\n",
    "blast_result_df = pd.DataFrame.from_records(hits_data)\n",
    "\n",
    "# If there are hits, normalize the \"Hits\" column\n",
    "if not blast_result_df.empty and 'Hits' in blast_result_df.columns:\n",
    "    df_hits_expanded = pd.json_normalize(blast_result_df['Hits'].explode(), sep='_')\n",
    "    blast_result_df = pd.concat([blast_result_df.drop(columns=['Hits']), df_hits_expanded], axis=1)\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "print(blast_result_df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge all results together. Take only the smallest e-value for BLAST results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\n",
      "g10513.t1             NaN\n",
      "g10514.t1             NaN\n",
      "g11320.t1             NaN\n",
      "g11513.t1    6.640000e-83\n",
      "g11806.t1    3.590000e-81\n",
      "g11960.t1    4.550000e-80\n",
      "g12388.t1    2.940000e-77\n",
      "g12510.t1    8.340000e-05\n",
      "g12562.t1    5.490000e-51\n",
      "g1285.t1     1.700000e-09\n",
      "g13530.t1    2.650000e-19\n",
      "g14472.t1    3.120000e-14\n",
      "g15153.t1             NaN\n",
      "g15484.t1    6.160000e-98\n",
      "g16318.t1    2.390000e-93\n",
      "g16368.t1    1.090000e-90\n",
      "g2203.t1     2.300000e-89\n",
      "g3428.t1     6.240000e-88\n",
      "g3679.t1     2.160000e-82\n",
      "g4106.t1     3.200000e-82\n",
      "g4970.t1     1.880000e-81\n",
      "g5237.t1     3.740000e-46\n",
      "g5443.t1     1.070000e-36\n",
      "g5467.t1     1.930000e-29\n",
      "g5502.t1     2.020000e-29\n",
      "g5503.t1     1.000000e-03\n",
      "g5510.t1     2.750000e-27\n",
      "g5616.t1     1.980000e-21\n",
      "g5641.t1     7.430000e-20\n",
      "g5927.t1     2.510000e-19\n",
      "g702.t1      6.070000e-19\n",
      "g7861.t1     4.800000e-02\n",
      "g8100.t1     7.790000e-19\n",
      "Name: E-value, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "\n",
    "df_min_evalue = df_result.groupby('Gene')['E-value'].min()\n",
    "\n",
    "print(df_min_evalue)\n",
    "# blast_result_df_top_eval = blast_result_df.groupby('Gene').\n",
    "# result_df = result_df.merge(blast_result_df_top_eval, how='inner', left_on='Gene', right_on='Gene')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
